### ISSUE TITLE
Match statement with a patterns referring to an out of scope identifier give incomplete warning

### ISSUE THREAD

**Issue by jbeeko** [dotnet/fsharp#6712]
Related #5475.

## What

The following code compiles with warnings which could be improved upon.

```
module DUs =
    type Countries = 
    | CA
    | US
    | GB
let x = DUs.CA

let output = 
    match x with 
    | US -> "US"
    | CA -> "CA"
    | _ -> "XX"
output
```

There is a warning on `| US -> "US"` that the value `US` is unused. 
<img width="327" alt="image" src="https://user-images.githubusercontent.com/4968704/57574078-22344000-73e7-11e9-9458-45a4878f03c8.png">

The default pattern has a warning that the rule will never be matched. 
<img width="385" alt="image" src="https://user-images.githubusercontent.com/4968704/57574081-2eb89880-73e7-11e9-9540-374ac0f1c8ad.png">

## Why

If when seeing the warnings you not taking into account that the pattern can be a value or a binding this is confusing in two ways. 
1. In the above example is is obvious that `US`  is out of scope. But in larger files this may not be obvious. Moreover normally referencing `US` without opening `DUs` would result in "The value or constructor US is not defined..." The message that it will never be used does not really point in the right direction.
2. The warning that the default pattern will never be matched is also confusing. Why not, should they not always match. And if the `CA` is never used why does the default not match?

And when this code is run it results in "US" which can also be confusing. 

Finally because there are only warnings this could result in a bug. 

## How

The compiler is assuming that `| CA` is a binding of x to CA not used on the right and emits a message to explain that situation. It should also suggest that CA might be an undefined value.  Something like

"Binding `CA` is not used in this rule or `CA` refers to a  value not defined..."

**Comment by jbeeko** [comment]
Whoops, I just noticed that when editing completed the match so my point about the default pattern having a warning is bit confusing. You actually get this warning even if the match is not complete. I edit the sample code to make this clear.

**Comment by isaacabraham** [comment]
The reason why the `_` default case (and `CA`) are never matched is because the `US` case is acting as the default case here i.e. all values are being bound to the symbol `US`. You could actually rewrite the first match clause in "full" like this: `| _ as US -> "US"` which is a more accurate representation of what is going on.

I agree that it's not intuitive though and could probably lead to bugs and runtime. I'm not sure what the solution to this would be though, especially without breaking changes.

**Comment by 0x53A** [comment]
Union cases already have a requirement to have an upper-case first character. Maybe do the reverse with bindings?

I'd propose to add a warning when a binding has an upper-case first character.

```f#
| US -> // warning: bindings in match cases should begin with a lower-case character
```

```f#
| us -> // no warning
```

**Comment by cartermp** [comment]
@0x53A That's a really good idea.

**Comment by jbeeko** [comment]
@0x53A @cartermp 

The existing error message already indicates that bindings should not be upper case. My point it that it assumes the issue is either a misspelled identifier or a contrary to convention an upper case binding name. Our team's experience is that this is often do to a missing open statement and the error message should also suggest that. 

See https://github.com/dotnet/fsharp/pull/8936

**Comment by abelbraaksma** [comment]
@jbeeko is correct, the warning suggested by @0x53A already exists. In VS, you see both warnings at the same time:

![image](https://user-images.githubusercontent.com/16015770/79114592-f0a8bf00-7d83-11ea-8f9e-65e23718b631.png)

But there's a bug with that warning. Consider the following, which doesn't raise FS0049:

![image](https://user-images.githubusercontent.com/16015770/79114813-86dce500-7d84-11ea-9a5c-5b53ff5561ea.png)

Versus:

![image](https://user-images.githubusercontent.com/16015770/79114840-96f4c480-7d84-11ea-9bde-d2e3d1642074.png)

Which does raise the FS0049 warning: 

> FS0049: Uppercase variable identifiers should not generally be used in patterns, and may indicate a misspelt pattern name.

In other words, that warning is only triggered if the the uppercase variable binding is three-or-more characters. I think that's a bug, or perhaps there's some historical reason why short names doesn't trigger it.

And I think it would certainly be a nice-to-have to add the "open module XXX" to the suggested quick-fixes:

![image](https://user-images.githubusercontent.com/16015770/79115071-2dc18100-7d85-11ea-990c-96251a56d3cb.png)

**Comment by BillHally** [comment]
@abelbraaksma said:
>In other words, that warning is only triggered if the the uppercase variable binding is three-or-more characters. I think that's a bug, or perhaps there's some historical reason why short names doesn't trigger it.

I've just looked at the history, and it seems it's been that way since the [fsharp repo was added to GitHub](https://github.com/dotnet/fsharp/blob/3900d411aeb2d7740f4e9fd72fce81b3a8f04357/src/fsharp/nameres.fs#L1892).

That file was renamed to [NameResolution.fs](https://github.com/dotnet/fsharp/blob/849401f0a6046d6b01f80293af109eafa51527bc/src/fsharp/NameResolution.fs#L3076) at some point, but that same line is still there.

There doesn't seem to be any explanation for why it's there.

**Comment by abelbraaksma** [comment]
Seems to be something that can be taken up by anybody and fixed?

**Comment by BillHally** [comment]
@abelbraaksma said:
> Seems to me to be something that can be taken up by anybody and fixed?

I guess that depends on whether the reason it's like that (whatever that is) still applies or not.

It certainly looks trivial to change.

**Comment by KathleenDollard** [comment]
That was a great example. It took me a few minutes to work out what was going on. 

Is this code pattern ever sane? And if it is not, is it recognizable:

```F#
    match x with 
    | y -> AnythingNotUsingY
    | AnythingOrDefayult -> Whatever
```

That is a pattern that is interpreted as a variable pattern that does not use the variable, followed by any pattern. Can we make this combination an error? Or make the first (a variable pattern that does not use the variable) a warning not just a "value is unused" info message. 

It's a wicked pothole.
